File: river/stats/kurtosis.py
from __future__ import annotations

from river import stats
from river.stats import _rust_stats


class Kurtosis(stats.base.Univariate):
    """Running kurtosis using Welford's algorithm.

    Parameters
    ----------
    bias
        If `False`, then the calculations are corrected for statistical bias.

    Examples
    --------

    >>> from river import stats
    >>> import scipy.stats
    >>> import numpy as np

    >>> np.random.seed(42)
    >>> X = np.random.normal(loc=0, scale=1, size=10)

    >>> kurtosis = stats.Kurtosis(bias=False)
    >>> for x in X:
    ...     print(kurtosis.update(x).get())
    -3.0
    -2.0
    -1.5
    1.4130027920707047
    0.15367976585756438
    0.46142633246812653
    -1.620647789230658
    -1.3540178492487054
    -1.2310268787102745
    -0.9490372374384453

    >>> for i in range(2, len(X)+1):
    ...     print(scipy.stats.kurtosis(X[:i], bias=False))
    -2.0
    -1.4999999999999998
    1.4130027920707082
    0.15367976585756082
    0.46142633246812403
    -1.620647789230658
    -1.3540178492487063
    -1.2310268787102738
    -0.9490372374384459

    >>> kurtosis = stats.Kurtosis(bias=True)
    >>> for x in X:
    ...     print(kurtosis.update(x).get())
    -3.0
    -2.0
    -1.5
    -1.011599627723906
    -0.9615800585356089
    -0.6989395431537853
    -1.4252699121794408
    -1.311437071070812
    -1.246289111322894
    -1.082283689864171

    >>> for i in range(2, len(X)+1):
    ...     print(scipy.stats.kurtosis(X[:i], bias=True))
    -2.0
    -1.4999999999999998
    -1.0115996277239057
    -0.9615800585356098
    -0.6989395431537861
    -1.425269912179441
    -1.3114370710708125
    -1.2462891113228936
    -1.0822836898641714

    References
    ----------
    [^1]: [Wikipedia article on algorithms for calculating variance](https://www.wikiwand.com/en/Algorithms_for_calculating_variance#/Covariance)

    """

    def __init__(self, bias=False):
        super().__init__()
        self.bias = bias
        self._kurtosis = _rust_stats.RsKurtosis(bias)

    def update(self, x):
        self._kurtosis.update(x)
        return self

    def get(self):
        return self._kurtosis.get()

File: rust_src/kurtosis.rs
use num::{Float, FromPrimitive};
use std::ops::{AddAssign, SubAssign};

use crate::moments::CentralMoments;
use crate::stats::Univariate;
use serde::{Deserialize, Serialize};
/// Running Kurtosis.
/// # Arguments
/// * `bias` - If `false`, then the calculations are corrected for statistical bias.
/// # Examples
/// ```
/// use river::kurtosis::Kurtosis;
/// use river::stats::Univariate;
/// let data: Vec<f64> = vec![ 0.49671415, -0.1382643 ,  0.64768854,  1.52302986, -0.23415337,-0.23413696];
/// let mut running_kurtosis: Kurtosis<f64> = Kurtosis::default();
/// for x in data.iter(){
///     running_kurtosis.update(*x);
///     println!("Kurtosis value: {}", running_kurtosis.get());
/// }
/// assert_eq!(running_kurtosis.get(), 0.46142635465045007);
/// ```
/// With bias enabled.
/// ```
/// use river::kurtosis::Kurtosis;
/// use river::stats::Univariate;
/// let data: Vec<f64> = vec![ 0.49671415, -0.1382643 ,  0.64768854,  1.52302986, -0.23415337,-0.23413696];
/// let mut running_kurtosis: Kurtosis<f64> = Kurtosis::new(true);
/// for x in data.iter(){
///     running_kurtosis.update(*x);
///     println!("Kurtosis with bias value: {}", running_kurtosis.get());
/// }
/// assert_eq!(running_kurtosis.get(), -0.6989395355484169);
/// ```
/// # References
/// [^1]: [Wikipedia article on algorithms for calculating variance](https://www.wikiwand.com/en/Algorithms_for_calculating_variance#/Covariance)
#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
pub struct Kurtosis<F: Float + FromPrimitive + AddAssign + SubAssign> {
    pub bias: bool,
    pub central_moments: CentralMoments<F>,
}
impl<F: Float + FromPrimitive + AddAssign + SubAssign> Kurtosis<F> {
    pub fn new(bias: bool) -> Self {
        Self {
            central_moments: CentralMoments::new(),
            bias,
        }
    }
}

impl<F> Default for Kurtosis<F>
where
    F: Float + FromPrimitive + AddAssign + SubAssign,
{
    fn default() -> Self {
        Self {
            central_moments: CentralMoments::new(),
            bias: false,
        }
    }
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Univariate<F> for Kurtosis<F> {
    #[inline(always)]
    fn update(&mut self, x: F) {
        self.central_moments.count.update(x);
        self.central_moments.update_delta(x);
        self.central_moments.update_m1(x);
        self.central_moments.update_sum_delta();
        self.central_moments.update_m4();
        self.central_moments.update_m3();
        self.central_moments.update_m2();
    }
    #[inline(always)]
    fn get(&self) -> F {
        let n = self.central_moments.count.get();
        let _0 = F::from_f64(0.).unwrap();
        let _1 = F::from_f64(1.).unwrap();
        let _2 = F::from_f64(2.).unwrap();
        let _3 = F::from_f64(3.).unwrap();

        let mut kurtosis: F = _0;
        if self.central_moments.m2 != _0 {
            let m2 = self.central_moments.m2;
            kurtosis = kurtosis + n * self.central_moments.m4
                / (m2 * m2);
        }
        if (!self.bias) && n > _3 {
            let nm1 = n - _1;
            return _1
                / (n - _2)
                / (n - _3)
                * ((n * n - _1) * kurtosis
                    - _3 * nm1 * nm1);
        }
        kurtosis - _3
    }
}
